### ISSUE TITLE
Issues with 2D bilinear least squares particle interpolation

### ISSUE THREAD

**Issue by naliboff** [geodynamics/aspect#3371]
I've started testing using active particles for lithospheric dynamics problems with sticky air and have run into quite a few issues the bilinear least squares method. 

One such issue is the assertion of the following error message during the early stages of a model run:
```
The violated condition was: 
    selected_properties.n_selected_components(n_particle_properties) == 1
Additional information: 
    Interpolation of multiple components is not supported.
```
I've attached the parameter file that produced this error, along with the associated log.txt and a file with the full error message.

I'm not doing adaptive refinement and the interpolation works fine on previous time steps, so at first glance I'm not sure why this error would be invoked. Using the 'cell average' interpolation works without issue (I ran it 30 time steps then stopped).

@sac-bsa - Have you encountered similar issues working with the 3D bilinear interpolation? 

Perhaps I've set the problem in a way that is incompatible with the bilinear interpolation, but there are no Asserts to catch the issue at the beginning?

The short-term (spring) goal is to rewrite the continental extension cookbook to show the tradeoffs between different interface tracking methods (particles, standard composition, DG, VOF) and techniques for speedup (averaging, AMG vs GMG, etc).

**Comment by naliboff** [comment]
Relevant files:
[log.txt](https://github.com/geodynamics/aspect/files/4206326/log.txt)

[error_message.txt](https://github.com/geodynamics/aspect/files/4206330/error_message.txt)

[test_2D_bilinear_4.prm.txt](https://github.com/geodynamics/aspect/files/4206332/test_2D_bilinear_4.prm.txt)

**Comment by sac-bsa** [comment]
When I had talked to @gassmoeller this summer about this assertion, he had mentioned something along the lines of aspect doesn't simultaneously work with multiple particle properties. The assertion that you ran into tests that idea, since the bilinear as implemented doesn't work for more than one particle property at a time. Cell average does work since it has a for loop that iterates over the particle properties. Since your model is able to invoke this aspect must now be interpolating multiple components in the same call to the interpolator. Although if my hypothesis is true, then I still don't understand how it makes it to the 11th timestep before it crashes. 
I figure there are at least two reasonable sounding ways to implement this functionality: for loop over the particle properties(like cell average) or changing the vectors r and c to become matrices. It would nice to hear @gassmoeller's advice on which might be the more desirable/correct way to implement this functionality.

**Comment by gassmoeller** [comment]
I think either implementation should work. If you go with the loop make sure to not reassemble the A matrix every time, but only loop over the bottom part of the function.

**Comment by egpuckett** [comment]
When we have implemented the loop for multiple particle properties we need to go back to the recent GJI paper and rerun some of the SolKz and SolCx computations that have both density and viscosity on the particles.

**Comment by gassmoeller** [comment]
@egpuckett: I had a quick look and I do not think it is necessary to reevaluate our results. The compositional fields are actually interpolated one by one at the moment (otherwise we also should have seen no convergence in our models at all, but all of our models nicely converged just at initially unexpected rates).
@naliboff: I think you ran into this problem because you tracked strain components on the particles and they are used for the viscosity somewhere, do you think this might be the reason? In other words your model uses functionality (interpolating several components in one call) that has so far not been used in other models. I agree we should support this in the bilinear least squares interpolator and implementing the loop that @sac-bsa and @egpuckett suggested would be good in either case, but I do not think any old results are wrong.

**Comment by naliboff** [comment]
@gassmoeller - that was my suspicion as well, but I obtained the same error (also at the same time step) after turning off strain weakening and only tracking the initial composition on the particles (see attached file). Running outside gdb the model crashed at the same time step, but within gdb it crashed one time step later and provided no stack. I'm not sure if the latter issue is related to something else on my side, but certainly odd behavior. 


[test_2D_bilinear_v2.prm.txt](https://github.com/geodynamics/aspect/files/4238409/test_2D_bilinear_v2.prm.txt)

**Comment by gassmoeller** [comment]
Ah, I found it. It is not dependent on the particle property, but on the fact that you have set a minimum number of particles per cell, and when a new particle is generated because there are too few per cell (apparently this first happens in timestep 11 of your model) then all particle properties from surrounding particles are interpolated to that new particle. Since this is what is not supported by the bilinear interpolator at the moment this is what crashes your model. The best fix is still to support this operation in the interpolator.

**Comment by sac-bsa** [comment]
@naliboff We have implemented a fix that allows your model to run to completion of the prm file you uploaded. Do you need this immediately? We are still concerned about checking some cases such as AMR, and creating a test.

**Comment by naliboff** [comment]
@sac-bsa - Great news! No rush from my end, I likely won't be able to get back to the associated models until next week. Thanks again.

**Comment by gassmoeller** [comment]
Closed by #3405. @naliboff: If you continue to see issues feel free to reopen.
